### PR TITLE
perf: improve genesis handling

### PR DESCRIPTION
**Pull request by DaniPopes**
- cache genesis hash and header (which includes state root) with `OnceLock` fields; these are currently computed at least 2 times each on each reth invocation, if not more (`sealed_genesis_header` computes header and hash which itself computes header again if not set)
- remove a ton of clones and btreemap collections in `impl HistoryWriter for DatabaseProvider` by passing required values in `impl IntoIterator`s
- pass state provider by reference in `init_genesis`; it's not really "cheap" to clone (5 arcs + `PruneModes` which could allocate)

### CODE REVIEW

File: crates/chainspec/src/spec.rs [hunk 1]
```diff
@@ -182,11 +180,14 @@ pub struct ChainSpec {
     ///
     /// This acts as a small cache for known chains. If the chain is known, then the genesis hash
     /// is also known ahead of time, and this will be `Some`.
-    pub genesis_hash: Option<B256>,
+    pub genesis_hash: OnceCell<B256>,
```
Comment (onbjerg) [inline crates/chainspec/src/spec.rs:183]: please update the docs as well bossman

**Review by onbjerg — DISMISSED** [review]

File: crates/chainspec/src/spec.rs [hunk 2]
```diff
@@ -275,7 +277,11 @@ impl ChainSpec {
     }
 
     /// Get the header for the genesis block.
-    pub fn genesis_header(&self) -> Header {
+    pub fn genesis_header(&self) -> &Header {
+        self.genesis_header.get_or_init(|| self.make_genesis_header())
```
Comment (mattsse) [inline crates/chainspec/src/spec.rs:281]: cool, this definitely works

**Review by mattsse — APPROVED** [review]
lgtm, pending @onbjerg nits

**Review by mattsse — APPROVED** [review]

**Comment by mattsse** [comment]
smol import conflict

**Review by onbjerg — APPROVED** [review]

**Pull request merged**